eras/conway/impl/src/Cardano/Ledger/Conway/Rules/HardFork.hs

```diff
@@ -102,8 +106,19 @@ updateDRepDelegations certState =
 
 populateVRFKeyHashes :: PState era -> PState era
 populateVRFKeyHashes pState =
-  let allVRFKeyHashes =
-        spsVrf
-          <$> Map.elems (pState ^. psStakePoolsL)
-            <> Map.elems (pState ^. psFutureStakePoolsL)
-   in pState & psVRFKeyHashesL .~ Set.fromList allVRFKeyHashes
+  pState
+    & psVRFKeyHashesL
+      %~ accumulateVRFKeyHashes (pState ^. psStakePoolsL)
+        . accumulateVRFKeyHashes (pState ^. psFutureStakePoolsL)
+  where
+    accumulateVRFKeyHashes ::
+      Map (KeyHash 'StakePool) StakePoolState ->
+      Map (VRFVerKeyHash 'StakePoolVRF) (NonZero Word64) ->
+      Map (VRFVerKeyHash 'StakePoolVRF) (NonZero Word64)
+    accumulateVRFKeyHashes spsMap acc =
+      Map.foldr' (\sps -> addVRFKeyHashOccurrence (sps ^. spsVrfL)) acc spsMap
+    addVRFKeyHashOccurrence vrfKeyHash =
+      Map.insertWith combine vrfKeyHash (knownNonZeroBounded @1)
+      where
+        -- Saturates at maxBound: if (+1) would overflow to 0, keep existing value
+        combine _ oldVal = fromMaybe oldVal $ mapNonZero (+ 1) oldVal
```

Comment:
Not sure if this is the best approach here. Would it be better to error in case of overflow?  Since thisi s called in Epoch , i thought it safer to do this instead of erroring.